TESTS/mbed_drivers/ticker/main.cpp

```diff
@@ -1,46 +1,73 @@
+ * Copyright (c) 2013-2017 ARM Limited
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may
- * not use this file except in compliance with the License.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
- * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
 
-
-/*
- * Tests is to measure the accuracy of Ticker over a period of time
- *
- *
- * 1) DUT would start to update callback_trigger_count every milli sec, in 2x callback we use 2 tickers
- *    to update the count alternatively.
- * 2) Host would query what is current count base_time, Device responds by the callback_trigger_count
- * 3) Host after waiting for measurement stretch. It will query for device time again final_time.
- * 4) Host computes the drift considering base_time, final_time, transport delay and measurement stretch
- * 5) Finally host send the results back to device pass/fail based on tolerance.
- * 6) More details on tests can be found in timing_drift_auto.py
- */
-
 #include "mbed.h"
 #include "greentea-client/test_env.h"
 #include "utest/utest.h"
 #include "unity/unity.h"
 
+
 using namespace utest::v1;
 
 #define ONE_MILLI_SEC 1000
 volatile uint32_t callback_trigger_count = 0;
 static const int test_timeout = 240;
 static const int total_ticks = 10;
 
+/* CPU clock */
+extern uint32_t SystemCoreClock;
+
+#define TOLERANCE_FACTOR 10000.0f
+#define US_FACTOR 1000000.0f
+
+/* Tolerance in microseconds
+
+   Assume below code
+
+   ...
+   Ticker ticker;
+   Timer timer;
+   int32_t ret;
+
+   timer.start();
+   ticker.attach_us(callback(wait_callback_sem), 100000);
+   ret = ticker_callback_sem.wait();
+   timer.stop();
+   int time_diff = timer.read_us();
+   TEST_ASSERT_UINT32_WITHIN((uint32_t)TOLERANCE, 100000, time_diff);
+   ...
+
+   For 100000[us] callback for NUCLEO_F070RB we get time_diff: 100070[us]
+   So the extra 70[us] is from time measure functions calls
+
+   To take in to account this extra time we introduces TOLERANCE value based on CPU clock (speed)
+
+   e.g.
+   For K64F          TOLERANCE = (10000 / 120000000) * 1000000 =  83[us]
+   For NUCLEO_F070RB TOLERANCE = (10000 /  48000000) * 1000000 = 208[us]
+   For NRF51_DK      TOLERANCE = (10000 /  16000000) * 1000000 = 625[us]
+ */
+#define TOLERANCE ((TOLERANCE_FACTOR / (float)SystemCoreClock) * US_FACTOR)
```

Comment:
What about setting the tolerance for the slowest device? It would make this simpler. Its good as you have it as well.

Reply:
The tolerance is scaled according CPU clock so it fits slow devices too. nrf51_dk is one of the slowest devices (form mbed compatible list).